data/campaigns/fri04.wmf/scripting/mission_thread.lua

```diff
@@ -0,0 +1,197 @@
+local seen_reebaud = false
+local seen_amazons = false
+
+local o_defeat_amz = nil
+local o_defeat_emp = nil
+local amazons_defeated_by_reebaud = false
+
+function hurry_up()
+   local did_warn = false
+   while true do
+      sleep(10000)
+      if seen_reebaud or p4.defeated then return end
+      if #p2:get_buildings("frisians_headquarters") < 1 then
+         if p2.defeated then
+            campaign_message_box(reebaud_defeated_1)
+            wl.ui.MapView():close()
+            return
+         elseif not did_warn then
+            campaign_message_box(reebaud_in_danger)
+            did_warn = true
+         end
+      end
+   end
+end
+
+function check_player_defeated(p, msg)
+   while true do
+      sleep(10000)
+      if p.defeated then
+         campaign_message_box(msg)
+         wl.ui.MapView():close()
+         return
+      end
+   end
+end
+
+function warn_empire_expansion()
+   while (
+         #p3:get_buildings("empire_blockhouse") +
+         #p3:get_buildings("empire_sentry") +
+         #p3:get_buildings("empire_outpost") +
+         #p3:get_buildings("empire_barrier") +
+         #p3:get_buildings("empire_tower") +
+         #p3:get_buildings("empire_fortress") +
+         #p3:get_buildings("empire_castle")) < 4 do
+      sleep(100000)
+      if seen_reebaud then return end
+   end
+   campaign_message_box(legate_expands)
+end
+
+function see_amazons(field)
+   if amazons_defeated_by_reebaud then return end
+   scroll_to_field(field)
+   sleep(2000)
+   if not seen_reebaud then
+      campaign_message_box(amz_1a)
+      campaign_message_box(amz_1b)
+   else
+      campaign_message_box(amz_1c)
+   end
+   sleep(500)
+   campaign_message_box(amz_2)
+   sleep(500)
+   campaign_message_box(amz_3)
+   if not o_defeat_amz then o_defeat_amz = add_campaign_objective(obj_defeat_amz) end
+   while not p4.defeated do sleep(1000) end
+   if seen_reebaud then campaign_message_box(victory_amz_a) else campaign_message_box(victory_amz_b) end
+   set_objective_done(o_defeat_amz)
+end
+
+function see_reebaud(field)
+   local empire_already_defeated = (p3.defeated or not campaign_data.payment)
```

Comment:
logically this variable is superfluous, as it is only dependent on whether we took the legate with us or not as attacking rules do forbid to attack the empire for everyone until this line.